sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,279 @@
+        let public_key_text = get_text_from_imds(&client, &url, &imds_session_token)?;
+        let public_key = public_key_text.trim_end();
+        // Simple check to see if the text is probably an ssh key.
+        if public_key.starts_with("ssh") {
+            debug!("{}", &public_key);
+            public_keys.push(public_key.to_string())
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid key. Skipping...",
+                &public_key_text
+            );
+            continue;
+        }
+    }
+    if public_keys.is_empty() {
+        warn!("No valid keys found");
+    }
+    Ok(public_keys)
+}
+
+/// Returns a list of public key urls strings for the public keys in IMDS. Since IMDS returns the
+/// list of available public keys as '0=my-public-key', we need to strip the index from the list and
+/// insert it into the key url.
+fn build_public_key_urls(public_key_list: &str) -> Vec<String> {
+    let mut public_key_urls = Vec::new();
+    for available_key in public_key_list.lines() {
+        let f: Vec<&str> = available_key.split('=').collect();
+        // If f[0] isn't a number, then it isn't a valid index.
+        if f[0].parse::<u32>().is_ok() {
+            let public_key_url = format!("{}/{}/openssh-key", IMDS_PUBLIC_KEY_BASE_URL, f[0]);
+            public_key_urls.push(public_key_url);
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid index. Skipping...",
+                &f[0]
+            );
+            continue;
+        }
+    }
+    if public_key_urls.is_empty() {
+        warn!("No valid key urls found");
+    }
+    public_key_urls
+}
+
+/// Store the args we receive on the command line.
+struct Args {
+    log_level: LevelFilter,
+}
+
+/// Print a usage message in the event a bad arg is passed
+fn usage() {
+    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
+    eprintln!(
+        r"Usage: {}
+            [ --log-level trace|debug|info|warn|error ]",
+        program_name
+    );
+}
+
+/// Parse the args to the program and return an Args struct
+fn parse_args(args: env::Args) -> Result<Args> {
+    let mut log_level = None;
+
+    let mut iter = args.skip(1);
+    while let Some(arg) = iter.next() {
+        match arg.as_ref() {
+            "--log-level" => {
+                let log_level_str = iter.next().ok_or_else(|| error::Error::Usage {
+                    message: "Did not give argument to --log-level".into(),
+                })?;
+                log_level = Some(LevelFilter::from_str(&log_level_str).map_err(|_| {
+                    error::Error::Usage {
+                        message: format!("Invalid log level '{}'", log_level_str),
+                    }
+                })?);
+            }
+
+            x => {
+                return Err(error::Error::Usage {
+                    message: format!("unexpected argument '{}'", x).into(),
+                })
```

Comment:
```suggestion
                return error::Usage {
                    message: format!("unexpected argument '{}'", x),
                }.fail()
```
A few things for this and similar spots.  With snafu, you probably shouldn't be using the `error::Error` enum directly, rather the structs that snafu generates based on them, hence the removal of `Error::`.  It gives you the `fail()` helper to handle the conversion and creation of `Err` around it.  Then, you shouldn't need to use `into()` when creating a snafu error type, since it does that for you.

This situation, where you need to return an error but can't use `ensure!`, is probably where snafu is the least helpful, but it's still best to fit to convention, and make use of snafu's better modes elsewhere.